Lib/inspect.py

```diff
@@ -792,25 +792,57 @@ def findsource(object):
         return lines, 0
 
     if isclass(object):
-        name = object.__name__
-        pat = re.compile(r'^(\s*)class\s*' + name + r'\b')
-        # make some effort to find the best matching class definition:
-        # use the one with the least indentation, which is the one
-        # that's most probably not inside a function definition.
-        candidates = []
-        for i in range(len(lines)):
-            match = pat.match(lines[i])
-            if match:
-                # if it's at toplevel, it's already the best one
-                if lines[i][0] == 'c':
-                    return lines, i
-                # else add whitespace to candidate list
-                candidates.append((match.group(1), i))
-        if candidates:
-            # this will sort by whitespace, and by line number,
-            # less whitespace first
-            candidates.sort()
-            return lines, candidates[0][1]
+        import ast
+
+        class ClassVisitor(ast.NodeVisitor):
+
+            def __init__(self, source, qualname, *args, **kwargs):
+                self.stack = []
+                self.source = source
+                self.line_number = None
+                self.qualname = qualname
+                super().__init__(*args, **kwargs)
+
+            def visit_ClassDef(self, node):
+                set_top_level_class = False
+
+                # Push the node on stack when there is a module level class
+                if not self.stack:
```

Comment:
Why only if the stack empty? What about deeper nesting?
```python
class A:
    class B:
        class C:
            pass
```

Reply:
I have refactored the code as per the comments. This was due to my stack management logic to add top level module and to push children class node inside the for loop. Since I have refactored the code to push and pop only at one place it's not needed. There is a similar test at https://github.com/python/cpython/pull/10307/files#diff-041789a617178e1a31ea7a1f86a5d2faR204 . Let me know if I need to add this also as a test.